### PR TITLE
[WIP] Improving UPDATE FROM Sql generation for PostgreSQL

### PR DESCRIPTION
**Pull request by sdanyliv**
Also improved emulation for other providers which do not support UPDATE FROM query syntax.
Let's check samples of the same linq query: https://github.com/linq2db/linq2db/blob/68276bce1cb39bdda828cfd4d8a41da5adb1ae86/Tests/Linq/Update/UpdateFromTests.cs#L131

### SqlServer 
```sql
UPDATE
	[t1]
SET
	[t1].[Value1] = ([t1].[Value1] * [t].[Value1]) * @int1,
	[t1].[Value2] = ([t1].[Value2] * [t].[Value2]) * @int2,
	[t1].[Value3] = ([t1].[Value3] * [t].[Value3]) * @int3
FROM
	[UpdatedEntities] [t1]
		INNER JOIN [NewEntities] [t] ON [t].[id] = [t1].[id]
WHERE
	[t].[id] <> @someId
```
### PostgreSql
```sql
UPDATE
	"UpdatedEntities"
SET
	"Value1" = ("UpdatedEntities"."Value1" * t."Value1") * :int1,
	"Value2" = ("UpdatedEntities"."Value2" * t."Value2") * :int2,
	"Value3" = ("UpdatedEntities"."Value3" * t."Value3") * :int3
FROM
	"NewEntities" t
WHERE
	t.id <> :someId AND t.id = "UpdatedEntities".id
```
### MySql, Access
```sql
UPDATE
	`UpdatedEntities` `t1`
		INNER JOIN `NewEntities` `t` ON `t`.`id` = `t1`.`id`
SET
	`t1`.`Value1` = (`t1`.`Value1` * `t`.`Value1`) * @int1,
	`t1`.`Value2` = (`t1`.`Value2` * `t`.`Value2`) * @int2,
	`t1`.`Value3` = (`t1`.`Value3` * `t`.`Value3`) * @int3
WHERE
	`t`.`id` <> @someId
```
### Providers which do not support UPDATE FROM syntax (SQLite, Oracle, ect.)
```sql
UPDATE
	[UpdatedEntities]
SET
	[Value1] = (
		SELECT
			([t1].[Value1] * [t].[Value1]) * @int1
		FROM
			[UpdatedEntities] [t1]
				INNER JOIN [NewEntities] [t] ON [t].[id] = [t1].[id]
		WHERE
			[t].[id] <> @someId AND [UpdatedEntities].[id] = [t1].[id]
	),
	[Value2] = (
		SELECT
			([t3].[Value2] * [t2].[Value2]) * @int2
		FROM
			[UpdatedEntities] [t3]
				INNER JOIN [NewEntities] [t2] ON [t2].[id] = [t3].[id]
		WHERE
			[t2].[id] <> @someId AND [UpdatedEntities].[id] = [t3].[id]
	),
	[Value3] = (
		SELECT
			([t5].[Value3] * [t4].[Value3]) * @int3
		FROM
			[UpdatedEntities] [t5]
				INNER JOIN [NewEntities] [t4] ON [t4].[id] = [t5].[id]
		WHERE
			[t4].[id] <> @someId AND [UpdatedEntities].[id] = [t5].[id]
	)
WHERE
	EXISTS(
		SELECT
			*
		FROM
			[UpdatedEntities] [t7]
				INNER JOIN [NewEntities] [t6] ON [t6].[id] = [t7].[id]
		WHERE
			[t6].[id] <> @someId AND [UpdatedEntities].[id] = [t7].[id]
	)
```

Remaining tasks:
- [x] Fix tests for other providers
- [x] Fix Merge tests

### CODE REVIEW

File: Tests/Tests.Playground/Tests.Playground.csproj [hunk 1]
```diff
@@ -14,6 +14,7 @@
     <Compile Include="..\Linq\Create\CreateData.cs" Link="CreateData.cs" />
     <Compile Include="..\Linq\Update\UpdateFromTests.cs" Link="UpdateFromTests.cs" />
     <Compile Include="..\Linq\Update\UpdateTests.cs" Link="UpdateTests.cs" />
+    <Compile Include="..\Linq\UserTests\SelectManyUpdateTests.cs" Link="SelectManyUpdateTests.cs" />
```
Comment (MaceWindu) [inline Tests/Tests.Playground/Tests.Playground.csproj]: you really should consider adding playground project to .gitignore :)

Comment (sdanyliv) [reply Tests/Tests.Playground/Tests.Playground.csproj]: Decided to do not do that. It is good to continue work from other place and have prepared testing project.

**Comment by AndrePScope** [comment]
Hi!

I think Oracle has more good candidate for such thing:
```sql
UPDATE
  [UpdatedEntities]
SET
  ( [Value1], [Value2], [Value3] ) = 
  (
    SELECT
      ([UpdatedEntities].[Value1] * [t].[Value1]) * @int1,
      ([UpdatedEntities].[Value2] * [t].[Value2]) * @int2,
      ([UpdatedEntities].[Value3] * [t].[Value3]) * @int3
    FROM
      [NewEntities] [t] 
    WHERE
      [t].[id] <> @someId AND [UpdatedEntities].[id] = [t].[id]
  )
WHERE
  EXISTS(
    SELECT
      *
    FROM
       [NewEntities] [t6] 
    WHERE
      [t6].[id] <> @someId AND [UpdatedEntities].[id] = [t6].[id]
  )
  ```

**Comment by BenoitStutz** [comment]
This Works on SQL-Server but not not on Oracle:
                var q = from w in objDbContext.W_DOWNLOAD_VTB
                        join b in objDbContext.VERSICHERTE_BETRIEB on new { w.BETRIEB_ID, w.VERSINT_ID } equals new { b.BETRIEB_ID, b.VERSINT_ID }
                        where w.REPORT_ID == lngReportId
                           && b.GUELTIG_AB == (from b2 in objDbContext.VERSICHERTE_BETRIEB
                                               where b2.VERSINT_ID == b.VERSINT_ID
                                                  && b2.BETRIEB_ID == b.BETRIEB_ID
                                               select b2.GUELTIG_AB).Max()
                        select new { w, b };

                q.Update(objDbContext.W_DOWNLOAD_VTB, obj => new Type
                {
                    EINTRITT_BETRIEB = obj.b.GUELTIG_AB
                });

**Comment by MaceWindu** [comment]
Could you open new issue?

**Pull request merged**